opencog/atomspace/AtomTable.cc

```diff
@@ -498,23 +495,45 @@ Handle AtomTable::add(AtomPtr atom, bool async)
     atom->keep_incoming_set();
     atom->setAtomSpace(_as);
```

Comment:
I think you have to set the atomspace under the lock, as otherwise, if the insert fails, you have this atom that advertises that its in an atomspace, but isn't actually in it.  Right?